### PR TITLE
grc: refactor c++ generation and fix zero-port busses

### LINKED ISSUE #7069: C++ generation with busports broken
### What happened?

#7067  fixes Python, but C++ doesn't work so far

### System Information

main v3.11.0.0git-632-g148a1091 (Python 3.11.6) 

### GNU Radio Version

3.11-git (main)

### Specific Version

v3.11.0.0git-632-g148a1091 (Python 3.11.6)

### Steps to Reproduce the Problem

Generation fails for:

cppgen_bus

https://gist.github.com/marcusmueller/5cd33ac2f6cf752049ddf7a50b7bc974

### Relevant log output

```shell
Generating: "/home/marcus/src/reproductions/cppgen/busports"
>>> Warning: This flow graph may not have flow control: no audio or RF hardware blocks found. Add a Misc->Throttle block to your flow graph to avoid CPU congestion.
Generate Error: 'CppTopBlockGenerator' object has no attribute 'get_parent_by_type'
>>> Failure
Traceback (most recent call last):
  File "/home/marcus/.usrlocal/lib64/python3.11/site-packages/gnuradio/grc/gui/Application.py", line 758, in _handle_action
    generator.write()
  File "/home/marcus/.usrlocal/lib64/python3.11/site-packages/gnuradio/grc/core/generator/cpp_top_block.py", line 101, in write
    for filename, data in self._build_cpp_header_code_from_template():
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/marcus/.usrlocal/lib64/python3.11/site-packages/gnuradio/grc/core/generator/cpp_top_block.py", line 162, in _build_cpp_header_code_from_template
    connections=self._connections(),
                ^^^^^^^^^^^^^^^^^^^
  File "/home/marcus/.usrlocal/lib64/python3.11/site-packages/gnuradio/grc/core/generator/cpp_top_block.py", line 495, in _connections
    connection = fg.parent_platform.Connection(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/marcus/.usrlocal/lib64/python3.11/site-packages/gnuradio/grc/gui/canvas/connection.py", line 39, in __init__
    super(self.__class__, self).__init__(*args, **kwargs)
  File "/home/marcus/.usrlocal/lib64/python3.11/site-packages/gnuradio/grc/core/Connection.py", line 56, in __init__
    param_factory = self.parent_platform.make_param
                    ^^^^^^^^^^^^^^^^^^^^
  File "/home/marcus/.usrlocal/lib64/python3.11/site-packages/gnuradio/grc/core/utils/descriptors/_lazy.py", line 48, in __get__
    value = self.func(instance)
            ^^^^^^^^^^^^^^^^^^^
  File "/home/marcus/.usrlocal/lib64/python3.11/site-packages/gnuradio/grc/core/base.py", line 108, in parent_platform
    return self.get_parent_by_type(Platform)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/marcus/.usrlocal/lib64/python3.11/site-packages/gnuradio/grc/core/base.py", line 103, in get_parent_by_type
    return parent.get_parent_by_type(cls)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'CppTopBlockGenerator' object has no attribute 'get_parent_by_type'
```

### PR DESCRIPTION
**Pull request by marcusmueller**
## Description

Busports flowgraphs couldn't be generated. Martin fixed the Python case, this fixes the C++ case.

As a drive-by, this untangles a bit of the block.py code, which was necessary for me to actually understand it.

## Related Issue

Fixes #7069

## Which blocks/areas does this affect?

GRC

## Testing Done

made https://gist.github.com/marcusmueller/5cd33ac2f6cf752049ddf7a50b7bc974 compile

## Checklist

- [x] I have read the [CONTRIBUTING document](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md).
- [x] I have squashed my commits to have one significant change per commit. 
- [x] I [have signed my commits before making this PR](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md#dco-signed)
- [x] My code follows the code style of this project. See [GREP1.md](https://github.com/gnuradio/greps/blob/main/grep-0001-coding-guidelines.md).
- [ ] I have updated [the documentation](https://wiki.gnuradio.org/index.php/Main_Page#Documentation) where necessary.
- [ ] I have added tests to cover my changes, and all previous tests pass.

### CODE REVIEW

**Comment by willcode** [comment]
Seems like too much to put into 3.10.9.2. Opinions?

**Comment by marcusmueller** [comment]
> Seems like too much to put into 3.10.9.2. Opinions?

Ha! I might be biased, and would opine otherwise :dagger: 

But let's look at that in detail: 

Commit [grc: fix C++ gen with busports](https://github.com/gnuradio/gnuradio/pull/7072/commits/57c6873816bc134f606a253a196fbefa15803ca6) fixes the C++ generation issue and completes #7067 

Commit [grc: minor refactoring of cpp_top_block](https://github.com/gnuradio/gnuradio/pull/7072/commits/6b2f9d7542992f7634b2f8e81c501dc8e9c15d8b) is but minor mechanical refactorings, which don't *need* to be backported, but would make `main` easier to work with and not really hurt maint-3.10

Commit [grc: fix busport creation with 0 ports](https://github.com/gnuradio/gnuradio/pull/7072/commits/b327c65b2f9561415e52331d9fd9924778e78183) is a bit "raw" in that it combines a bugfix (which should be backported, but isn't very high prio) with refactoring. That's mostly due to me having to do the refactoring to be able to fix the bug.

I can look into extracting the quality-of-life improvements out of that commit (that is, mainly, consistent string quoting/f-strings), but it'd be sad to also remove the added logging, which was helpful knowing where things went wrong. The actual refactoring was rather benign, in essence: extract a function and convert 

```python
def handle_ports(self, ports):
  for thing in {"foo", "bar"}:
      if thing == "foo":
       baz = self.a[thing]      
     elif thing == "bar":
       baz = self.b[thing]
     a_lot_of_code_here_using_ports_and_baz
```

to 

```python
def do_thing(self, baz):
     a_lot_of_code_here_using_ports_and_baz

def handle_ports(self, ports):
    if not ports:
      return_early
    self.do_thing(ports, self.a["foo"])
    self.do_thing(ports, self.b["bar"])
```

**Comment by willcode** [comment]
I'll give it a closer look. I was hoping we could keep the 3.10.9.2 changes down to really small things that are easily checked in a diff.

**Comment by willcode** [comment]
Amended last commit with python formatting, cleaned up commit message.

File: grc/core/generator/top_block.py [hunk 1]
```diff
@@ -13,7 +10,7 @@
 
 DATA_DIR = os.path.dirname(__file__)
 
-PYTHON_TEMPLATE = os.path.join(DATA_DIR, 'flow_graph.py.mako')
```
Comment (willcode) [inline grc/core/generator/top_block.py:16]: Why change all `'` to `"`?

**Comment by mbr0wn** [comment]
Can you point out the bugfix for 0-length bus ports? The ' to " conversion makes this diff hard to read.

File: grc/core/generator/cpp_top_block.py [hunk 2]
```diff
@@ -492,10 +506,10 @@ def by_domain_and_blocks(c):
                         for port_num in porta.bus_structure:
                             hidden_porta = porta.parent.sources[port_num]
                             hidden_portb = portb.parent.sinks[port_num]
-                            connection = fg.parent_platform.Connection(
```
Comment (willcode) [inline grc/core/generator/cpp_top_block.py:495]: This is the line equivalent to the Python fix.

File: grc/core/blocks/block.py [hunk 3]
```diff
@@ -150,70 +152,76 @@ def rekey(ports):
...
+                        removed_bus_ports.append(port)
+                        removed_bus_connections.append(c)
+                ports.remove(port)
+        return (removed_bus_ports, removed_bus_connections)
+
+    def _handle_direction(self, direc: str, ports: list, bus_state: bool) -> None:
+        assert direc in {'sink', 'source'}
+        if not ports:
+            self.current_bus_structure[direc] = None
+            return
```
Comment (marcusmueller) [inline grc/core/blocks/block.py:171]: This is the core fix, @mbr0wn

**Comment by marcusmueller** [comment]
> Can you point out the bugfix for 0-length bus ports? The ' to " conversion makes this diff hard to read.

sure, @mbr0wn , https://github.com/gnuradio/gnuradio/pull/7072#discussion_r1446234792 ; but find that in the "from" code…

**Comment by willcode** [comment]
Split simple fix commit out to another PR for quick merge.

We need to back out the unneeded `'` -> `"` changes. I'm also curious what the "busport creation with 0 ports" means.

**Comment by willcode** [comment]
Do portions of this still apply after the related PR that was merged?

**Comment by marcusmueller** [comment]
@willcode good question, don't have resources right now to look at that in detail

> I'm also curious what the "busport creation with 0 ports" means.

You could right-click on a block with no output ports, and select "create bus source port", and it would suddenly get a (very confusing, and dysfunctional) white bus output port, combining all 0 ports into 1 bus port. (same for sink)

**Pull request closed without merging**